playground/src/server/cache.ts

```diff
@@ -0,0 +1,64 @@
+/**
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type {WasmResponse} from './wasmFunctions';
+
+// A cache of URL and HTTP response, and a function to get all URLs visited by
+// this cache.
+interface RecordedCache {
+  get: (url: string) => Promise<WasmResponse>;
+  put: (url: string, response: WasmResponse) => Promise<void>;
+  visitedUrls: () => Set<string>;
+}
+
+// An in-memory cache to store URLs and their HTTP response.
+export class SubresourceCache {
```

Comment:
(opt) Consider simplifying this as `export class SubresourceCache implements RecordedCache`, inlining the recording code directly into this class's methods. On the other hand, if you're planning in a future PR to extract `createRecorder` as its own public wrapper class with multiple backing cache classes, then splitting the caching/recording functionality makes sense.

Reply:
The reason caching/recording is splitted is because the in-memory cache is shared by all tested SXGs but the recorded list is owned by each SXG.

If we want `SubresourceCache` to implement `RecordedCache`, we will have three options
1. No longer share in-memory cache across all SXGs, and re-fetch subresources for each SXG.
1. Use a static global hash map as in-memory cache.
1. Create a `InMemoryCache` class, and construct `new SubresourceCache(inMemoryCache)` for each SXG.

Reply:
Ah, I missed that detail. OK, it LGTM as it is.